apps/shipments/permissions.py

```diff
@@ -12,6 +14,31 @@ def has_object_permission(self, request, view, obj):
         return obj.owner_id == request.user.id
 
 
+class IsShipmentOwner(permissions.BasePermission):
+    """
+    Custom permission to only allow owners of an object to edit it
+    """
+
+    def has_permission(self, request, view):
+        shipment = Shipment.objects.get(pk=view.kwargs['shipment_pk'])
+
+        # Permissions are only allowed to the owner of the shipment.
+        return shipment.owner_id == request.user.id
+
+
+class IsSharedOrOwner(permissions.BasePermission):
+    def has_object_permission(self, request, view, obj):
+        if obj.owner_id == request.user.id:
+            return True
+        if 'permission_link' in request.query_params and request.query_params['permission_link']:
+            permission_link = PermissionLink.objects.get(pk=request.query_params['permission_link'])
```

Comment:
This line will throw an exception if the PermissionLink does not exist.  Ultimately it's presented to the user as a 404 on the Shipment, which in itself isn't a terrible thing.  However, it's not immediately obvious from the code that this is the line generating the 404.  I recommend wrapping this line in a Try/Catch and returning `False` if a ObjectDoesNotExist exception is thrown.  This will change the return on the Shipment to a 403 Forbidden which is acceptable in this case.  (this will affect the unit test)